Scripts/static-analyzer/Tests/AnalyzerTests/SourceNavigatorTests.swift

```diff
@@ -0,0 +1,193 @@
+import XCTest
+import IndexStoreDB
+
+@testable import Analyzer
+
+class SourceNavigatorTests: XCTestCase {
+
+    func testLookupSymbols() throws {
+        let support = TestSupport()
+
+        let code = """
+            import Foundation
+            func foo<T>(a: Int, b: Bool, c: String, d: T) -> NSArray {
+                fatalError("Unimplemented")
+            }
+            func bar() {
+                _ = foo(a: 0, b: false, c: "", d: 1.0)
+            }
+            """
+        let navigator = try support.navigator(forSourceCode: code)
+
+        let symbols = navigator.lookupSymbols(name: "foo(a:b:c:d:)", in: support.sourceFile)
+        XCTAssertEqual(symbols.count, 1)
+
+        let symbol = try XCTUnwrap(symbols.first)
+        let callSites = navigator.callSites(of: USR(rawValue: symbol.usr)!)
+        XCTAssertEqual(callSites.count, 1)
+    }
+
+    func testLookupExpression() throws {
+        let support = TestSupport()
+
+        let code = """
+            import Foundation
+            func foo<T>(a: Int, b: Bool, c: String, d: T) -> NSArray {
+                fatalError("Unimplemented")
+            }
+            func bar() {
+                _ = foo(a: 0, b: false, c: "", d: 1.0)
+            }
+            """
+        let navigator = try support.navigator(forSourceCode: code)
+
+        let symbols = navigator.lookupSymbols(name: "foo(a:b:c:d:)", in: support.sourceFile)
+        XCTAssertEqual(symbols.count, 1)
+
+        let symbol = try XCTUnwrap(symbols.first)
+        let callSites = navigator.callSites(of: USR(rawValue: symbol.usr)!)
+        XCTAssertEqual(callSites.count, 1)
+
+        let location = SymbolLocation(path: support.sourceFile.pathString, moduleName: "", line: 6, utf8Column: 10)
+        let expressions = try navigator.expressions(at: location).map { $0.type }
+        XCTAssertTrue(expressions.contains("(Int, Bool, String, Double) -> NSArray"))
+    }
+
+    func testLookupInstanceMethod() throws {
+        let support = TestSupport()
+        let code = """
+            protocol ProtocolA {
+                func map(_ transform: Any) -> Any
+            }
+            func foo(p: ProtocolA) {
+              _ = p.map(0)
+            }
+            func bar(p: ProtocolA) {
+              _ = p.map("")
+            }
+            func baz() {
+              _ = [1, 2, 3].map { $0 * 2}
+            }
+            """
+        let navigator = try support.navigator(forSourceCode: code)
+
+        let proto = try navigator.resolveType(named: "ProtocolA")
+
+        let mapFuncs = navigator.lookupInstanceMethods(named: "map(_:)", of: proto)
+        XCTAssertEqual(mapFuncs.count, 1)
+        let mapFunc = try XCTUnwrap(mapFuncs.first)
+
+        let mapCallSites = navigator.callSites(of: mapFunc)
+        XCTAssertEqual(mapCallSites.count, 2)
+        let lines = mapCallSites.map { $0.line }
+        XCTAssertEqual(Set(lines), [5, 8])
```

Comment:
I don't think I ever stopped and thought about line numbers starting from 1 and not 0... 😅